docs/rfcs/consistent-timeout-handling.md

```diff
@@ -0,0 +1,405 @@
+  MinTimeout time.Duration `mapstructure:"min_timeout"
+}
+```
+
+The two fields are meant to be interpreted in combination, according to
+this matrix:
+
+| Explanation | Timeout=0 | Timeout>0 |
+| -- | -- | -- |
+| **MinTimeout<0** | No minimum, no maximum case. In this case, the deadline is explicitly erased. | A maximum timeout is imposed, but the deadline is not checked. |
+| **MinTimeout≥0** | The deadline is checked, but a new maximum is not imposed. Allows existing and no-deadline requests to pass. | The deadline is checked and a maximum is imposed. |
+
+The current component uses a default Timeout of `5s`. This proposal
+would use default MinTimeout of `0`, meaning to enforce a non-negative
+deadline, rejecting expired contexts. Proposed context-handling logic is listed below.
+
+### Retry sender
+
+This component is already deadline-aware.
+
+### Queue sender
+
+A new field will be introduced, with default matching the
+original behavior of this component.
+
+```golang
+  // FailFast indicates that the queue should immediately
+  // reject requests when the queue is full, without considering
+  // the request deadline. Default: true.
+  FailFast bool `mapstructure:"fail_fast"`
+```
+
+In case the new FailFast flag is false, there are two cases:
+
+1. The request has a deadline. In this case, wait until the deadline
+   to enqueue the request.
+2. The request has no deadline. In this case, let the request block
+   indefinitely until it can be enqueued.
+
+### Batch processor/sender
+
+As discussed in the companion RFC, batching processes should
+transmit errors, which means waiting for batch responses, and
+they should respect cancellation.
+
+This proposal states that batching processes should also use
+an outgoing context deadline. Batching logic should keep track
+of the maximum deadline over requests with a deadline in each
+batch. The batch deadline shall be set to the maximum deadline
+of any in the batch. Only when all requests in the batch do not
+have a deadline, will the outgoing batch request have no
+request deadline.
+
+Under this scheme, requests with no deadline, combined into a batch
+of requests that do have a deadline will cause the deadline-free request
+to have a deadline. This is intentional. Users are encouraged
+to use deadlines always and/or use separate pipelines for requests
+with and without deadlines.
+
+### Receiver helper
+
+Receivers are the entry point for data entering a Collector
+pipeline. Receivers are responsible for setting the initial
+context deadline for each request. This proposal states that
+receivers should be encouraged with an opt-in mechanism to
+offer standard timeout configuration. The configuration of
+this opt-in support follows the Timeout sender:
+
+- `min_timeout` (duration): Limits the allowable timeout for new requests to a minimum value. >=0 means deadline checking.
+- `timeout` (duration): Limits the allowable timeout for new requests to a maximum value. Must be >= 0.
```

Comment:
Why not having a processor for this?

Reply:
I support the idea of a processor to disregard or reset the timeout, but I also think it's just as good done in the timeout sender.

I've stated that I think all the functionality of the exporterhelper would be useful in a processor, I'd call that a "pipeline processor". https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/35803#issuecomment-2593778577